reactos/drivers/usb/usbport/queue.c

```diff
@@ -0,0 +1,1388 @@
+
+            KeReleaseSpinLock(&Endpoint->EndpointSpinLock,
+                              Endpoint->EndpointOldIrql);
+            return;
+        }
+
+        Entry = Transfer->TransferLink.Flink;
+    }
+
+    Entry = Endpoint->TransferList.Flink;
+
+    while (Entry && Entry != &Endpoint->TransferList)
+    {
+        Transfer = CONTAINING_RECORD(Entry,
+                                     USBPORT_TRANSFER,
+                                     TransferLink);
+
+        if (Transfer->Flags & TRANSFER_FLAG_ABORTED)
+        {
+            DPRINT_CORE("USBPORT_FlushAbortList: Aborted ActiveTransfer - %p\n",
+                        Transfer);
+
+            KeReleaseSpinLock(&Endpoint->EndpointSpinLock,
+                              Endpoint->EndpointOldIrql);
+            return;
+        }
+
+        Entry = Transfer->TransferLink.Flink;
+    }
+
+    AbortList = &Endpoint->AbortList;
+
+    while (!IsListEmpty(AbortList))
+    {
+        //DbgBreakPoint();
+
+        Irp = CONTAINING_RECORD(AbortList->Flink,
+                                IRP,
+                                Tail.Overlay.ListEntry);
+
+        RemoveHeadList(AbortList);
+        InsertTailList(&List, &Irp->Tail.Overlay.ListEntry);
+    }
+
+    KeReleaseSpinLock(&Endpoint->EndpointSpinLock, Endpoint->EndpointOldIrql);
+
+    while (!IsListEmpty(&List))
+    {
+        //DbgBreakPoint();
+
+        Irp = CONTAINING_RECORD(List.Flink,
+                                IRP,
+                                Tail.Overlay.ListEntry);
+
+        RemoveHeadList(&List);
+
+        Urb = URB_FROM_IRP(Irp);
+
+        DeviceHandle = Urb->UrbHeader.UsbdDeviceHandle;
+        InterlockedDecrement(&DeviceHandle->DeviceHandleLock);
+
+        Status = USBPORT_USBDStatusToNtStatus(Urb, USBD_STATUS_SUCCESS);
+
+        DPRINT_CORE("USBPORT_FlushAbortList: complete Irp - %p\n", Irp);
+
+        Irp->IoStatus.Status = Status;
+        Irp->IoStatus.Information = 0;
+        IoCompleteRequest(Irp, IO_NO_INCREMENT);
+    }
+}
+
+VOID
+NTAPI
+USBPORT_FlushCancelList(IN PUSBPORT_ENDPOINT Endpoint)
+{
+    PDEVICE_OBJECT FdoDevice;
+    PUSBPORT_TRANSFER Transfer;
+    PIRP Irp;
+    KIRQL OldIrql;
+    PUSBPORT_DEVICE_EXTENSION  FdoExtension;
+
+    DPRINT_CORE("USBPORT_FlushCancelList: ... \n");
+
+    FdoDevice = Endpoint->FdoDevice;
+    FdoExtension = FdoDevice->DeviceExtension;
+
+    KeAcquireSpinLock(&FdoExtension->FlushTransferSpinLock, &OldIrql);
+    KeAcquireSpinLockAtDpcLevel(&Endpoint->EndpointSpinLock);
+
+    while (!IsListEmpty(&Endpoint->CancelList))
+    {
+        Transfer = CONTAINING_RECORD(Endpoint->CancelList.Flink,
+                                     USBPORT_TRANSFER,
+                                     TransferLink);
+
+        RemoveHeadList(&Endpoint->CancelList);
+
+        Irp = Transfer->Irp;
+
+        if (Irp)
+        {
+            DPRINT("USBPORT_FlushCancelList: Irp - %p\n", Irp);
+
+            IoAcquireCancelSpinLock(&OldIrql);
```

Comment:
This overwrites your previous OldIrql and will cause this routine not to restore the irql. Needs a second variable